src/collapse/sample.rs

```diff
@@ -27,81 +31,108 @@ static START_LINE: &str = "Call graph:";
 // We know we're done when we get to this line.
 static END_LINE: &str = "Total number in stack";
 
-/// Settings that change how frames are named from the incoming stack traces.
-///
-/// All options default to off.
-#[derive(Clone, Debug, Default)]
+/// `sample` folder configuration options.
+#[derive(Clone, Debug)]
 pub struct Options {
-    /// Don't include modules with function names
+    /// Don't include modules with function names. Default is `false`.
     pub no_modules: bool,
+
+    /// The number of threads to use. Default is the number of logical cores on your machine.
+    pub nthreads: usize,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            no_modules: false,
+            nthreads: *common::DEFAULT_NTHREADS,
+        }
+    }
 }
 
 /// A stack collapser for the output of `sample` on macOS.
 ///
 /// To construct one, either use `sample::Folder::default()` or create an [`Options`] and use
 /// `sample::Folder::from(options)`.
-#[derive(Default)]
+#[derive(Clone)]
 pub struct Folder {
+    /// Number of samples for the current stack frame.
+    current_samples: usize,
+
+    /// The number of stacks per job to send to the threadpool.
+    nstacks_per_job: usize,
+
     /// Function on the stack in this entry thus far.
     stack: Vec<String>,
 
-    /// Number of samples for the current stack frame.
-    current_samples: usize,
     opt: Options,
 }
 
-impl Collapse for Folder {
-    fn collapse<R, W>(&mut self, mut reader: R, mut writer: W) -> io::Result<()>
+impl Default for Folder {
+    fn default() -> Self {
+        Self {
+            current_samples: 0,
+            nstacks_per_job: DEFAULT_NSTACKS_PER_JOB,
+            stack: Vec::default(),
+            opt: Options::default(),
+        }
+    }
+}
+
+impl CollapsePrivate for Folder {
```

Comment:
I don't think this implementation of `CollapsePrivate` is going to work correctly for `sample::Folder` when using multiple threads. In the section of a `sample` profile below, this collapser will detect the end of a stack on the second to last line. If it splits there, the chunk given to the first thread will be fine because it has all of the stack frames, but the second thread will get its input starting already indented multiple levels into the stack. It won't know what functions were below it in the stack.

You could split it whenever you find a new stack root, but that doesn't work with the interface of `CollapsePrivate` since it's based on detecting the end of each stack. It's probably not even worth doing though. Instead, I think you should remove the requirement of `Collapse` to be multi-threaded and make this implementation single-threaded. 

```
    825 Thread_15758535
    + 825 thread_start  (in libsystem_pthread.dylib) + 13  [0x7fff739c240d]
    +   825 _pthread_start  (in libsystem_pthread.dylib) + 66  [0x7fff739c6249]
    +     825 _pthread_body  (in libsystem_pthread.dylib) + 126  [0x7fff739c32eb]
    +       825 std::sys::unix::thread::Thread::new::thread_start::h8b5ec703435993e3  (in rg) + 9  [0x103a106c9]
    +         825 std::sys_common::thread::start_thread::he4c2173d5991c786  (in rg) + 136  [0x103a20eb8]
    +           825 _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::h8612a2a83552fc2d  (in rg) + 165  [0x10383e155]
    +             825 __rust_maybe_catch_panic  (in rg) + 31  [0x103a230cf]
    +               825 std::panicking::try::do_call::h08ca64c12e8a9bb0 (.llvm.15516475458848885060)  (in rg) + 37  [0x103837885]
    +                 825 std::sys_common::backtrace::__rust_begin_short_backtrace::h71644e3bf9e2bb80  (in rg) + 37  [0x10384d7f5]
    +                   794 ignore::walk::Worker::run::h428457710ae7645a  (in rg) + 1035  [0x10393278b]
    +                   ! 794 rg::search_parallel::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h6e849b55a66fcd85  (in rg) + 172  [0x1037cbcac]
    +                   !   794 _$LT$rg..search..SearchWorker$LT$W$GT$$GT$::search_impl::hcf61284e58f410e0 (.llvm.17269878581919215071)  (in rg) + 5509  [0x1037ab3b5]
    +                   !     618 grep_searcher::searcher::Searcher::search_path::h4867309d69f27419  (in rg) + 127  [0x103749bdf]
    +                   !     : 618 std::fs::OpenOptions::_open::h1cd1c180d181e2f5  (in rg) + 33  [0x103a1e481]
    +                   !     :   618 std::sys::unix::fs::File::open::hb90e1c1c787080f0  (in rg) + 198  [0x103a14ad6]
    +                   !     :     617 std::sys::unix::fs::File::open_c::h3660b29577d73852  (in rg) + 338  [0x103a14c72]
    +                   !     :     | 617 std::sys::unix::cvt_r::h0e253dc028f09451  (in rg) + 53  [0x103a1ac55]
    +                   !     :     |   617 __open  (in libsystem_kernel.dylib) + 10  [0x7fff739051ee]
    +                   !     :     1 std::sys::unix::fs::File::open_c::h3660b29577d73852  (in rg) + 372  [0x103a14c94]
    +                   !     119 grep_searcher::searcher::Searcher::search_path::h4867309d69f27419  (in rg) + 2905  [0x10374a6b9]

```

Reply:
Yeah, I think it's probably fine for `Sample` to remain single-threaded for now.

Reply:
👍